feat(config): add Docker DNS hostname resolution support

gustavovalverde opened the pull request:
## Summary

This PR adds support for Docker DNS hostname resolution in validator address configuration, allowing addresses like `zebra:18232` instead of requiring IP addresses. It also makes cookie authentication config-based rather than address-type-based.

### Changes

- **Add hostname resolution support**: Changed `validator_jsonrpc_listen_address` and related fields from `SocketAddr` to `String` to accept hostnames
- **Add `zaino-common::net` module**: New utility module with `resolve_socket_addr()` and `try_resolve_address()` functions that handle both IP:port and hostname:port formats
- **Add format validation**: Invalid address formats (missing port, garbage input) fail immediately at config time with descriptive errors, while valid hostnames that can't be resolved (Docker DNS) are deferred to connection time
- **Config-based cookie auth**: Cookie authentication is determined solely by whether `validator_cookie_path` is configured, not by address type. This gives operators control over their security choices in trusted network environments.
- **Update config conversions**: Changed `From<ZainodConfig>` to `TryFrom<ZainodConfig>` for `StateServiceConfig` and `FetchServiceConfig` to properly handle fallible address resolution

### Validator Address Validation

The validation rules are:
- **Resolved IPs**: Must be private (RFC1918/ULA)
- **Hostnames**: Validated at connection time (supports Docker/K8s service discovery)
- **Cookie auth**: Determined by `validator_cookie_path` config, not enforced by address type

The `AddressResolution` enum distinguishes between:
- `Resolved(SocketAddr)` - Successfully resolved, private IP validation applies
- `UnresolvedHostname` - Valid hostname:port format but DNS lookup failed (acceptable for Docker)
- `InvalidFormat` - Malformed input that fails immediately

### Rationale for Cookie Auth Change

Previously, cookie authentication was forced for non-loopback IPs regardless of configuration. This created friction for operators in trusted network environments (Docker, Kubernetes, internal networks) where cookie auth adds operational complexity without proportional security benefit.

The change makes behavior consistent: if you configure `validator_cookie_path`, cookie auth is used; if you don't, it's not required. This respects the operator's explicit configuration choice while maintaining the private IP requirement as the network security boundary.

### Testing

- Added unit tests for address format validation (DNS-independent)
- Added tests for config-based cookie auth behavior:
  - `test_cookie_auth_not_forced_for_non_loopback_ip` - validates non-loopback private IPs work without cookie auth
  - `test_public_ip_still_rejected` - validates private IP requirement still applies
- Marked DNS-dependent tests with `#[ignore]` to prevent CI flakiness
- All 14 config tests pass

### Breaking Changes

- `test_node_and_return_url()` now takes `&str` instead of `SocketAddr`
- Config conversion traits changed from `From` to `TryFrom`
- Non-loopback private IPs no longer require cookie authentication (operator's choice)

gustavovalverde committed:
feat(config): add Docker DNS hostname resolution support

- Add `zaino-common::net` module with `resolve_socket_addr()` and
  `try_resolve_address()` for hostname:port resolution
- Change validator address config fields from `SocketAddr` to `String`
- Add `AddressResolution` enum to distinguish between resolved addresses,
  valid hostnames pending DNS, and invalid formats
- Reject malformed addresses (missing port, invalid format) at config time
- Defer DNS resolution for valid hostnames (Docker DNS) to connection time
- Update `From<ZainodConfig>` to `TryFrom<ZainodConfig>` for proper error handling
- Mark DNS-dependent tests with `#[ignore]` to prevent CI flakiness

zancas commented on zainod/src/config.rs:
```diff
@@ -139,16 +170,6 @@ impl ZainodConfig {
                     "TLS required when connecting to external addresses.".to_string(),
                 ));
             }
-
```
@nachog00 @AloeareV opinions?

This seems like a separable concern?  Something that could be its own PR?

gustavovalverde replied:
You're right that the cookie auth change is a separable concern conceptually. However, it's directly related to this PR because of a behavior inconsistency it exposed:

Before this PR, the code forced cookie auth for non-loopback IPs. With hostname support added, we had a drift:
- Resolved IPs → cookie auth forced (address-type-based)
- Hostnames → no enforcement (config-based)

This inconsistency meant the same logical configuration behaved differently depending on whether you used `192.168.1.10:18232` vs `zebra:18232`.

To fix this I made both behave the same way - cookie auth is now determined solely by whether `validator_cookie_path` is configured. This:
1. Respects operator configuration choices in trusted networks
2. Makes behavior consistent across address formats
3. Keeps the private IP requirement as the security boundary

zancas reviewed (commented):
I looked this over briefly.  Why is cookie auth for non-locahost checking removed in this PR?  Isn't it a separate concern to DNS host resolution?

gustavovalverde committed:
refactor(config): make cookie auth config-based, not address-type-based

Remove forced cookie authentication requirement for non-loopback IPs.
Cookie auth is now determined solely by whether validator_cookie_path
is configured, giving operators control over their security choices
in trusted network environments.

- Remove is_loopback_listen_addr function (no longer needed)
- Simplify hostname resolution log message
- Add tests for symmetric behavior across address types

nachog00 committed:
Merge branch 'dev' into feat/docker-hostname-resolution
